### PR TITLE
feat(ertp): Upgrade quotes to drop newly optional recovery sets

### LINKED ISSUE #8400: price-feed vats hold old QuotePayments in recovery set, causing storage leak
## Describe the bug

I spent a few days investigating storage usage on mainnet, and learned that about 60% of our kvstore entries are the result of a bug in the price feed contracts, and this count grows by about 10 rows for every new price quote created.

External oracles submit messages (signed txns) into the chain with claims about the e.g. ATOM/USD price at a given time. These get executed as contract operations (create an invitation with args that include the price claim, make an offer, get a seat, etc). The price feed contract then publishes a price quote to subscribers, such as an aggregator that averages/etc quotes from multiple sources to drive Vault creation or liquidation.

While the authority to produce a quote is limited by virtue of the wallets which have access to the price feed's objects, the quote objects themselves might be delivered to arbitrary subscribers (over untrusted paths), and those subscribers want to verify that the quote object they received is genuine. The current implementation uses ERTP `Payment` objects to represent the price quotes, and subscribers can call `E(priceQuoteIssuer).getAmountOf(allegedQuote)` to confirm their authenticity.

generating a quote object:
https://github.com/Agoric/agoric-sdk/blob/92b6cd72484079b0349d8ccfa4510aeb820e8d67/packages/zoe/src/contracts/priceAggregator.js#L103-L109

verifying one:
https://github.com/Agoric/agoric-sdk/blob/92b6cd72484079b0349d8ccfa4510aeb820e8d67/packages/zoe/src/contractSupport/priceAuthorityTransform.js#L90-L92

However, subscribers like the `scaledPriceAuthority` stop there: the quote payment is dropped, without being deposited or burned.

This interacts badly with a safety feature of ERTP: the Recovery Set. To guard against assets being lost, each time you call `payment = await E(purse).withdraw()`, the payment is added to the Purse's "recovery set": a strong SetStore. If you send the Payment to a correspondent, but they fail to deposit it, you can call [`getRecoverySet()`](https://github.com/Agoric/agoric-sdk/blob/92b6cd72484079b0349d8ccfa4510aeb820e8d67/packages/ERTP/src/purse.js#L91) to access all these Payments, and then deposit them back into a Purse. Or, you can call [`recoverAll()`](https://github.com/Agoric/agoric-sdk/blob/92b6cd72484079b0349d8ccfa4510aeb820e8d67/packages/ERTP/src/purse.js#L94) to sweep them back into the original Purse. Brand new tokens, created by `E(mint).mintPayment()`, are held in a special `mintRecoveryPurse`, available from the IssuerKit. All Payments are removed from their Recovery Set as soon as they are deposited in some other Purse, or burned.

Recovery normally isn't necessary, because well-written code will deposit a Payment as soon as it receives it, and only some sort of error would prevent that from happening. However, quotes-as-payments aren't assets to be hoarded: the code only uses ERTP Payments for their authenticity-verification properties, not their exclusive-transfer properties. So it is not surprising that the price contracts fail to `deposit()` or `burn()` them.

The consequence is that the Payment remains in the recovery set forever.

The contract and ERTP code uses a variety of SetStores, WeakMapStores, and Exos to express this logic. Swingset (specifically liveslots) supports those with "virtual collections" and "virtual objects", all of which bottoms out in a per-vat key-value table named the "vatstore". This holds rows to represent collection entries, ordinals to enable consistent iteration through collections, reference counts, weak-key reference-graph edges, export statuses, and finally the virtual-object state itself.

The particular data structures used by ERTP means that each Payment costs about 10 vatstore entries. In the following diagram, each circle is a vatstore entry, of various types.

v29 quote payment

And this is an example of one particular (still-extant) payment, where each gray box is a separate vatstore entry:

v29 quote payment example

Since each vatstore entry is really stored in the kernel's `kvStore` (a table inside the "SwingStore" SQLite DB), they cost 10 kvStore entries each. And because state-sync integrity requires us to mirror all kvStore items into the IAVL database, they also cost 10 IAVL entries each. We've been creating quotes once every five minutes since the Vaults feature was released last June, and by now we have about 120k quote payments kept alive in v29 (the ATOM-USD price-feed vat). A similar problem is keeping 90k payments alive in v46 (scaledPriceAuthority). As we add new denominations in the future, their price feed vats will have the same issue. I estimate this currently costs 440MB of space (combining both swingstore and IAVL), growing at 5MB/day.

## Additional Costs

v46 (scaledPriceAuthority) has 90k quote payments in this state, however it has also exported about 30k of them to v7-board. That vat is still holding weak references (c-list entries in the "recognizable" state, as opposed to "reachable"). This causes additional kvstore entries for each one: two for the export-side c-list, two in the kernel object table, two in the import-side clist, and some amount of RAM in v7-board.

We believe that, at some point, v46 received the payment and sent it to `E(board.readOnlyMarshaller).toCapData()`. The serialization process submits the received quote-payment Presence object to `passStyleOf`, which would have concluded that the Presence was `passable`, and added the object to it's WeakMap cache. This WeakMap is virtualized by liveslots (to prevent userspace from sensing GC, as usual), and `VirtualObjectAwareWeakMap` remembers a real Map keyed by vref strings. So v7-board will spend some (small) amount of RAM on each such object. These WeakMap references cause the c-list to remain in the "recognizable" state: they do not represent a strong ("reachable") reference.

Both the c-list/kvstore entries and the RAM costs will go away if/when the quote payment objects are released.

We've been considering allowing the Endo/`passStyleOf` code to use a real WeakMap instead of the virtualized one. This requires the permission of liveslots (otherwise arbitrary user code could do it, violating our GC-sensing rules), so coordination between Swingset and Endo is ongoing. If implemented, this would remove the separate `passStyleOf` cache and would probably allow v9-board to avoid spending either RAM or c-list imports on the object.

## Potential Fixes

To flatten this growth rate, at today's Zoe meeting we discussed one short-term fix named "burn after reading": clients should burn their quote payment objects after verifying them. For example, the v29 price-feed quotes are received by v46-scaledPriceAuthority, where [`priceAuthorityTransform.js`](https://github.com/Agoric/agoric-sdk/blob/92b6cd72484079b0349d8ccfa4510aeb820e8d67/packages/zoe/src/contractSupport/priceAuthorityTransform.js#L90-L92), could do:
```js
    const { value: sourceQuoteValue } = await E(sourceQuoteIssuer).getAmountOf(
      sourceQuotePayment,
    );
   await E(sourceQuoteIssuer).burn(sourceQuotePayment); // burn after reading
```

We would need similar changes in all other places that receive quote payments. In particular, the scaled/aggregated quotes *produced* by v46-scaledPriceAuthority are delivered to v48-vaultFactory, where something needs its own `burn()` call.

Another approach would be to have the **producing** code rotate through recovery purses every hour or so:
* mint Payments as before, but immediately deposit that first payment into a separate recovery Purse, then `.withdraw()` a new one from that purse and send the *second* payment to the client
* then, once an hour, drop the old purse and start using a new one, relying upon GC to delete the recovery set, and thus forget about all the old payments (if these represented real assets, this would lose funds, but quotes aren't worth anything)
* note that this doesn't help remediate the existing payments, because those are held by the recovery set associated with the Issuer's single `mintRecoveryPurse`, which is closed over by `mintPayment()` and cannot be changed or destroyed short of a vat upgrade which first modifies `baggage` to remove the reference

In the longer term, we should find a lighter-weight means to verify the authenticity of price quotes. ERTP Payments offer both verification (through `issuer.getAmountOf()`) and exclusive transfer (through `purse.deposit(payment)`), but this use case does not require exclusive transfer. The ocap community provides other tools for mere verification, like the Branding pattern. The "issuer" replacement can retain a WeakMap of quote objects, and offer an API to perform `validQuotes.has(allegedQuote)`.

## Remediation

In addition to fixing new price quotes, we also need to dispose of the hundreds of thousands of existing ones. One idea from today's meeting:

* periodically (perhaps once per price quote), the producer could access `mintRecoveryPurse`, fetch its recovery set, iterate through a few dozen payments, and `burn` them all
  * to avoid invalidating a quote before a consumer gets a chance to verify it, check the timestamp on the quote, and only burn it if it's at least an hour older than the new quote being created
  * this would spread the work out over time: if we burn ten old payments upon each new 5-minute quote, we'd remediate all 120k from v29 in about 41 days, without causing an undue amount of work for any new quote

Another approach would be to have an upgraded producer vat immediately perform `mintRecoveryPurse.recoverAll()` (perhaps assuming that no quote payments are outstanding at the moment of upgrade, or accepting one quote getting invalidated *before* reading). However we need to be cognizant of the size of the recovery set: this would cause v29 to immediately delete about a million items from its vatstore, requiring maybe ten million syscalls, all in a single crank. Who knows how long that might take. In addition, that would remove a million rows from the IAVL tree, and we know that IAVL pruning takes a long time when there has been a lot of churn, so each validator might experience a significant (vote-missing) delay some arbitrary number of blocks after the upgrade (depending upon their individual pruning configuration).

Note that different approaches require upgrades to different vats, and other bugs make some vats are more practical to  upgrade than others. In particular, we'd like to fix the price-feed code before making a lot of new deployments, to avoid adding more load to the system, but eventually we need to fix (and also remediate) the existing vats.

cc @Chris-Hibbert @erights

### PR DESCRIPTION
**Pull request by erights**
closes: https://github.com/Agoric/agoric-sdk/issues/8400
refs: https://github.com/Agoric/agoric-sdk/issues/8417 https://github.com/Agoric/agoric-sdk/issues/8404 https://github.com/Agoric/agoric-sdk/issues/8401

## Description

Take 2 of https://github.com/Agoric/agoric-sdk/pull/8406

Make recovery sets optional.

Enable an old issuer with recovery sets to be upgraded to one without, that also empties and suppresses any recovery sets it inherits.

Upgrade quote issuers to empty and suppress their recovery sets. This is 
* appropriate for quote issuers because quote "payments" do not actually transfer tradable value. A new quote can always be obtained that is at least as good as an old lost quote.
* needed, because many existing quotes have been otherwise dropped, but are being pointlessly retained only by recovery sets.

One of the solution strategies for #8400 previously discussed is to deposit or burn the old quote "payments" in the recovery sets, in order to clean them up. This would be somewhat more traumatic for clients, though tolerably so. A client holding a pre-upgrade quote "payment" would find that it had been burned, requiring them to get a new payment if they wished to continue holding a valid one. This PR is much less traumatic. By just dropping the old recovery sets, any old payments that are actually being held remain as valid as they were. The only trauma would be anyone counting on recovering payments from recovery sets. We are confident no code using quotes does so.

### Security Considerations

Recovery sets are a safety feature. This PR introduces the option to suppress that safety feature, including by upgrading old issuers with that feature to one without. If such an upgrade could be done maliciously, clients depending on this safety feature could be surprised by its absence.

### Scaling Considerations

This PR is motivated by one scaling consideration: The storage leak of retaining old useless payments. In particular, dropped quote "payments".

If merged before #8417 is fixed, this PR would also cause a painful scaling issue. On upgrade, it will likely make a massive number of quote "payments" unreachable all at once. Currently, this will cause a large amount of gc-driven cleanup activity during that one crank and block, which can be problematic. Instead, the PR is written assuming #8417 has already been fixed, and depends on it to spread out the resulting gc-driven cleanup activity over a number of future cranks and blocks. Thus, the PR should not be merged until after #8417 is fixed.

Separate from this PR, we should find those places where we drop live payments of any kind, and see if we can safely burn or deposit them before dropping them.

### Documentation Considerations

Any documentation of issuer creation should document the new option for suppressing recovery sets. This should explain both the storage-leak hazard of not suppressing them and the safety hazard of suppressing them.

Any documentation on use of recovery sets to recover old payments should make clear that this feature is now optional, and some issuers, such as the quote issuers, omit the feature.

### Testing Considerations

The most important tests needed are the upgrade tests, discussed in the next section.

Even aside from upgrade, this PR needs more tests before it can become ready for review. At the time of this writing, all that is tested is that this PR does not break any existing tests. 
- [ ] We also need tests that create issuerKits without recovery sets, and test that these behave as expected.

Since the point of opting out of upgrade tests is to avoid a durable storage leak, we should have 
- [ ] a test where durable storage consumption grows linearly in the number of created and dropped payments with recovery sets, 
- [ ] and (after gc) maintains a steady state without recovery sets. 

Need to find out how to write such a test.

### Upgrade Considerations

Need to test 
- [ ] that old quote issuers with recovery sets upgrade to new quote issuers without recovery sets, 
- [ ] that otherwise they continue to pass all the same tests
- [ ] that their behavior now reflects the suppression of recovery sets
- [ ] that the old space for dropped payments is indeed recovered
- [ ] Once #8417 is fixed, that the work of recovering that storage is adequately spread out across cranks and blocks.

We should also test
- [ ] that an old issuer without recovery sets cannot (yet?) be upgraded into an issuer with recovery sets. That is the intent as of this PR. Eager for feedback whether this upgrade direction should be supported. We could, but I'd rather not if no one would use it.

### CODE REVIEW

File: packages/ERTP/src/issuerKit.js
Comment (erights) [inline packages/ERTP/src/issuerKit.js:118]: Won't work because `get` throws on absence. Must test with `has` instead.

(Hoist on own petard?)

File: packages/zoe/src/contractSupport/priceAuthorityQuoteMint.js
Comment (erights) [inline packages/zoe/src/contractSupport/priceAuthorityQuoteMint.js:22]: Oops! I missed that this uses a different baggage key name.

Must repair.

File: packages/ERTP/src/issuerKit.js
Comment (Chris-Hibbert) [inline packages/ERTP/src/issuerKit.js]:
```suggestion
 * What _should_ have been named `prepareIssuerKit`. Used to either revive an
 * ancestor issuer kit, or to make a new durable one if it is absent, and to place it
 * in baggage for the next successor.
```

File: packages/ERTP/src/types-ambient.js
Comment (Chris-Hibbert) [inline packages/ERTP/src/types-ambient.js]:
```suggestion
 * At this time, a `'noRecoverySets'` predecessor cannot be upgraded to a
```

File: packages/ERTP/src/paymentLedger.js
Comment (Chris-Hibbert) [inline packages/ERTP/src/paymentLedger.js:174]: Add something saying that issuers can choose to have no recovery sets, and probably to indicate that this makes sense for "payments" that don't represent the transfer of value, like price-quote payments.

Comment (erights) [reply packages/ERTP/src/paymentLedger.js:174]: Done

File: packages/ERTP/src/types-ambient.js
Comment (erights) [reply packages/ERTP/src/types-ambient.js]: Done

File: packages/ERTP/src/issuerKit.js
Comment (erights) [reply packages/ERTP/src/issuerKit.js]: Done

File: packages/ERTP/src/paymentLedger.js
Comment (Chris-Hibbert) [inline packages/ERTP/src/paymentLedger.js]:
```suggestion
   * transferable value. Recovery sets preserve payments that are otherwise
```

File: packages/zoe/src/zoeService/makeInvitation.js
Comment (Chris-Hibbert) [inline packages/zoe/src/zoeService/makeInvitation.js:25]: missing declaration of `q`

**Review by Chris-Hibbert — COMMENTED** [review]
I think we shouldn't merge this until it includes a test showing that the upgrade will be successful. If you like, you can assign that to the userland team (No, I don't know how to assign a PR to a team.)

If addressing the leak is urgent, talk to P.M. about prioritization, or we'll be glad to give advice about adding upgrade tests (`packages/deployment/upgrade-test/`)

File: packages/zoe/src/zoeService/makeInvitation.js
Comment (erights) [reply packages/zoe/src/zoeService/makeInvitation.js:25]: Done

File: packages/ERTP/src/paymentLedger.js
Comment (erights) [reply packages/ERTP/src/paymentLedger.js]: Done

**Comment by erights** [comment]
> I think we shouldn't merge this until it includes a test showing that the upgrade will be successful. If you like, you can assign that to the userland team (No, I don't know how to assign a PR to a team.)

Thanks for the offer! I agree, so I've put the PR in its current state into Draft. Let's revisit the mechanics of that reassignment once this PR otherwise seems to work.

> If addressing the leak is urgent, talk to P.M. about prioritization, or we'll be glad to give advice about adding upgrade tests (`packages/deployment/upgrade-test/`)

This testing could proceed as soon as this PR is otherwise green. But merging must wait until after SwingSet rate limits gc activity anyway, so this one could only become urgent after it is unblocked on that.

cc @warner @ivanlei @rowgraus @turadg

File: packages/zoe/src/contractSupport/priceAuthorityQuoteMint.js
Comment (erights) [reply packages/zoe/src/contractSupport/priceAuthorityQuoteMint.js:22]: Nevermind. Looking closer, it is only a name for internally tagging the baggage, which it uses to print itself and in error diagnostics. As far as the baggage tree, the old quote only uses `'quoteMintIssuerBaggage'`, so the new code is correct. 

There is an observable difference though: The provide* and prepare* convention uses one name for both purposes, so we lose the separate `'scaledPriceAuthority quoteMintIssuerBaggage'` for baggage born using this new PR. Fortunately, this change caused no tests to fail, so we expect it is inconsequential.

Resolving.

**Review by turadg — CHANGES_REQUESTED** [review]
Requesting that the PR description use [the pull request template](https://github.com/Agoric/agoric-sdk/blob/7710838df55e651ca2c8777950df5b3a3bba9059/.github/PULL_REQUEST_TEMPLATE.md). That has a section about Upgrade Considerations, under which should be mentioned the need for an upgrade test. 

Also requesting that this PR have such a test before merging. Userland can help with that, adapting https://github.com/Agoric/agoric-sdk/blob/83b1fc89eb892c9783cc79e0f5753a076ff3a494/packages/deployment/upgrade-test/upgrade-test-scripts/agoric-upgrade-12/zoe-full-upgrade/zcf-upgrade-script.js#L16

**Comment by erights** [comment]
> Requesting that the PR description use [the pull request template](https://github.com/Agoric/agoric-sdk/blob/7710838df55e651ca2c8777950df5b3a3bba9059/.github/PULL_REQUEST_TEMPLATE.md). That has a section about Upgrade Considerations, under which should be mentioned the need for an upgrade test.

Thanks for suggesting that! Done.


> Also requesting that this PR have such a test before merging. Userland can help with that, adapting
> 
> https://github.com/Agoric/agoric-sdk/blob/83b1fc89eb892c9783cc79e0f5753a076ff3a494/packages/deployment/upgrade-test/upgrade-test-scripts/agoric-upgrade-12/zoe-full-upgrade/zcf-upgrade-script.js#L16

I would really appreciate that help, thanks!

**Pull request closed without merging**
